src/providers/arcgisrest/qgsarcgisrestutils.cpp

```diff
@@ -0,0 +1,591 @@
+  if ( esriFieldType == "esriFieldTypeString" )
+    return QVariant::String;
+  if ( esriFieldType == "esriFieldTypeDate" )
+    return QVariant::Date;
+  if ( esriFieldType == "esriFieldTypeGeometry" )
+    return QVariant::Invalid; // Geometry column should not appear as field
+  if ( esriFieldType == "esriFieldTypeOID" )
+    return QVariant::LongLong;
+  if ( esriFieldType == "esriFieldTypeBlob" )
+    return QVariant::ByteArray;
+  if ( esriFieldType == "esriFieldTypeGlobalID" )
+    return QVariant::String;
+  if ( esriFieldType == "esriFieldTypeRaster" )
+    return QVariant::ByteArray;
+  if ( esriFieldType == "esriFieldTypeGUID" )
+    return QVariant::String;
+  if ( esriFieldType == "esriFieldTypeXML" )
+    return QVariant::String;
+  return QVariant::Invalid;
+}
+
+QgsWKBTypes::Type QgsArcGisRestUtils::mapEsriGeometryType( const QString& esriGeometryType )
+{
+  // http://resources.arcgis.com/en/help/arcobjects-cpp/componenthelp/index.html#//000w0000001p000000
+  if ( esriGeometryType == "esriGeometryNull" )
+    return QgsWKBTypes::Unknown;
+  else if ( esriGeometryType == "esriGeometryPoint" )
+    return QgsWKBTypes::Point;
+  else if ( esriGeometryType == "esriGeometryMultipoint" )
+    return QgsWKBTypes::MultiPoint;
+  else if ( esriGeometryType == "esriGeometryPolyline" )
+    return QgsWKBTypes::MultiCurve;
+  else if ( esriGeometryType == "esriGeometryPolygon" )
+    return QgsWKBTypes::Polygon;
+  else if ( esriGeometryType == "esriGeometryEnvelope" )
+    return QgsWKBTypes::Polygon;
+  // Unsupported (either by qgis, or format unspecified by the specification)
+  //  esriGeometryCircularArc
+  //  esriGeometryEllipticArc
+  //  esriGeometryBezier3Curve
+  //  esriGeometryPath
+  //  esriGeometryRing
+  //  esriGeometryLine
+  //  esriGeometryAny
+  //  esriGeometryMultiPatch
+  //  esriGeometryTriangleStrip
+  //  esriGeometryTriangleFan
+  //  esriGeometryRay
+  //  esriGeometrySphere
+  //  esriGeometryTriangles
+  //  esriGeometryBag
+  return QgsWKBTypes::Unknown;
+}
+
+static QgsPointV2* parsePoint( const QVariantList& coordList, QgsWKBTypes::Type pointType )
+{
+  int nCoords = coordList.size();
+  if ( nCoords < 2 )
+    return 0;
+  bool xok = false, yok = false;
+  double x = coordList[0].toDouble( &xok );
+  double y = coordList[1].toDouble( &yok );
+  if ( !xok || !yok )
+    return 0;
+  double z = nCoords >= 3 ? coordList[2].toDouble() : 0;
+  double m = nCoords >= 4 ? coordList[3].toDouble() : 0;
+  return new QgsPointV2( pointType, x, y, z, m );
+}
+
+static QgsCircularStringV2* parseCircularString( const QVariantMap& curveData, QgsWKBTypes::Type pointType, const QgsPointV2& startPoint )
+{
+  QVariantList coordsList = curveData["c"].toList();
+  if ( coordsList.isEmpty() )
+    return 0;
+  QList<QgsPointV2> points;
+  points.append( startPoint );
+  foreach ( const QVariant& coordData, coordsList )
+  {
+    QgsPointV2* point = parsePoint( coordData.toList(), pointType );
+    if ( points.last() == 0 )
```

Comment:
@manisandro this looks very strange to me. What's this line trying to do? At the moment it will compare the last added point and if its x coordinate is 0 it will return no circular string..